source/agora/test/Ledger.d

```diff
@@ -229,3 +229,58 @@ unittest
         assert(merkle_path.length == 0);
     }
 }
+
+/// test behavior of receiving double-spend transactions
+unittest
+{
+    import agora.common.Deserializer;
+    import agora.common.Serializer;
+    import std.algorithm;
+    import std.range;
+    import core.time;
+    import core.thread;
+    import vibe.core.log;
+
+    const NodeCount = 2;
+    auto network = makeTestNetwork(NetworkTopology.Simple, NodeCount, true,
+        100, 20, 100);  // reduce timeout to 100 msecs
+    network.start();
+    scope(exit) network.shutdown();
+
+    auto nodes = network.apis.values;
+    auto node_1 = nodes[0];
+
+    // we don't care about node 2
+    //nodes[1 .. $].each!(node => node.filter!(node.putTransaction));
+
+    auto txs = makeChainedTransactions(getGenesisKeyPair(), null, 1);
+    txs.each!(tx => node_1.putTransaction(tx));
+    containSameBlocks(nodes, 1).retryFor(3.seconds);
+
+    txs = makeChainedTransactions(getGenesisKeyPair(), txs, 1);
+    txs.each!(tx => node_1.putTransaction(tx));
+    containSameBlocks(nodes, 2).retryFor(3.seconds);
+
+    txs = makeChainedTransactions(getGenesisKeyPair(), txs, 1);
+
+    setLogLevel(LogLevel.info);
+
+    // create a deep-copy of the first tx
+    auto backup_tx = deserialize!Transaction(serializeFull(txs[0]));
+
+    // create a double-spend tx
+    txs[0].inputs[0] = txs[1].inputs[0];
+    txs[0].outputs[0].value = Amount(100);
+    auto signature = getGenesisKeyPair().secret.sign(hashFull(txs[0])[]);
+    txs[0].inputs[0].signature = signature;
```

Comment:
I'd add a check that the tx is valid after this

Reply:
Ah, that's a good idea!